File: java/killer-sudoku-helper/src/main/java/KillerSudokuHelper.java
import java.util.ArrayList; import java.util.List;

public class KillerSudokuHelper {

    List<List<Integer>> combinationsInCage(Integer cageSum, Integer cageSize, List<Integer> exclude) {
        List<List<Integer>> possibleCombinations = combinationsInCage(cageSum, cageSize);

        return possibleCombinations.stream()
                .filter(combination -> !combination.stream().anyMatch(number -> exclude.contains(number)))
                .toList();
    }

    List<List<Integer>> combinationsInCage(Integer cageSum, Integer cageSize) {
        List<List<Integer>> combinationsList = new ArrayList<>();
        combinationsInCage(combinationsList, new ArrayList<>(), cageSum, cageSize);
        return combinationsList;
    }
    
    void combinationsInCage(List<List<Integer>> combinations, List<Integer> currentList, Integer targetSum, int cageSize) {
        int currentSize = currentList.size();
        if (targetSum == 0 && currentSize == cageSize) {
            combinations.add(new ArrayList<>(currentList));
            return;
        }

        if (targetSum <= 0 || currentSize > cageSize) {
            return;
        }

        int first = currentSize > 0 ? currentList.get(currentSize - 1) + 1 : 1;
        for (int i = first; i <= targetSum; i++) {
            currentList.add(i);
            combinationsInCage(combinations, currentList, targetSum - i, cageSize);
            currentList.remove(currentList.size() - 1);
        }

    }

}

File: elixir/killer-sudoku-helper/lib/killer_sudoku_helper.ex
defmodule KillerSudokuHelper do
  @doc """
  Return the possible combinations of `size` distinct numbers from 1-9 excluding `exclude` that sum up to `sum`.
  """
  @spec combinations(cage :: %{exclude: [integer], size: integer, sum: integer}) :: [[integer]]
  def combinations(cage) do
    %{exclude: exclude, size: size, sum: sum} = cage

    combinations_in_cage([], [], sum, size)
    |> Enum.reject(fn list -> list |> Enum.any?(&(&1 in exclude)) end)
    |> Enum.reverse()
  end

  defp combinations_in_cage(combinations_list, current_list, target_sum, size)
       when target_sum <= 0 do
    if target_sum == 0 and current_list |> length() == size do
      [current_list |> Enum.reverse() | combinations_list]
    else
      combinations_list
    end
  end

  defp combinations_in_cage(combinations_list, current_list, target_sum, size) do
    first = if current_list |> Enum.empty?(), do: 1, else: (current_list |> hd()) + 1

    first..9
    |> Enum.reduce(combinations_list, fn i, acc ->
      combinations_in_cage(acc, [i | current_list], target_sum - i, size)
    end)
  end
end
